Refactor Plugin system

TobiasWalle opened the pull request:
Implementation of #577

tipsy commented:
> I left the configuration methods for backward compatibility.

I haven't looked at the PR yet, but you don't have to care about backward compatibility. The master branch is currently on 3.X, which isn't out yet.

TobiasWalle commented:
@tipsy Oh great. But at the moment I leave it at it is, so it is easier to understand.

tipsy commented:
This looks really good @TobiasWalle, we should definitely do something like this.

TobiasWalle commented:
@tipsy Great! What do you think would be the next steps? I think we should split the PR for the implementation from the refactoring of the existing code.

I think we could then also remove the old `register` function, as both solution solve the same problem, but the plugin system is in my opinion more powerful.

tipsy commented:
Yup, that's exactly how I think it should be done too.

PR1: Introduce plugin system with tests, remove `register`
PR2: Refactor existing code to use plugin system

dherges commented on src/main/java/io/javalin/core/Plugin.kt:
```diff
@@ -0,0 +1,25 @@
+ * that will occur in the application.
+ */
+@FunctionalInterface
+interface Plugin {
```
Hi @TobiasWalle, sounds good to me!

Do you think it's possible to change the implememtation of `Extension` and make it behave like the plugin system?

I think for users, it will be less confusing to have just one API to do the same thing, though in two slightly different fashions.

TobiasWalle replied:
That is true. If you are okay with that, I would remove the `Extension` in favor of the Plugins and update the doc comment.

I thought about just extending the `Extension` but as the name `plugin` is used a lot in the javalin packages, I think to name this concept `Plugin` would be a better fit.

dherges replied:
the two differences between `Extension` and `Plugin` come due to the addition of `JavalinConfig`. When I wrote the extensions the, `JavalinConfig` did not exist yet. The ideas of both are very similar and I think it's reasonable to merge extensions and plugins into one.

I'd move `Extension` here and add `ExtensionLifecycleInit` - wdyt?

TobiasWalle replied:
Sounds good. But I think we should either
* rename the the package `io.javalin.plugin` to `io.javalin.extension`
* rename `Extension` to `Plugin`

I think otherwise it can be confusing to the user if the `OpenApiExtension` is in the `plugin` package, as both words have a very similar meaning.

@tipsy Do you have any thoughts on the naming?

dherges replied:
> rename `Extension` to `Plugin`

sounds good to me. I took over the names from [Sinatra extensions](http://sinatrarb.com/extensions.html#extending-the-dsl-class-context-with-sinatraregister). I'm ok with deprecating extensions.

How many people are out there using the current `Extension` APIs? Do you want to print deprecation warnings or remove them instantly?

tipsy replied:
The current master is version 3.0, so we'll just remove them.

TobiasWalle replied:
Perfect. I will update the PR today or tomorrow.

tipsy commented:
I prefer plugin.

TobiasWalle commented:
@tipsy @dherges I updated the PR
* I removed the refactoring
* I added more tests, improved the api and some error handling
* I removed the extentions in favor of plugins. All use cases of plugins should be covered.

If I missed something, please let me know.

tipsy commented on src/main/java/io/javalin/core/JavalinConfig.java:
```diff
@@ -83,6 +93,34 @@
+    /**
+     * Returns true if the plugin is registered.
+     */
+    public boolean isPluginRegistered(@NotNull Class<? extends Plugin> pluginClass) {
```
Is this method required? Could you instead do 

```if (inner.plugins.containsKey(plugin.getClass())) {```

?

tipsy commented on src/main/java/io/javalin/core/JavalinConfig.java:
```diff
@@ -83,6 +93,34 @@
+    }
+
+    /**
+     * Get a registered plugin or null if it is not defined
```
```suggestion
     * Get a registered plugin (or null if it is not defined)
```

TobiasWalle replied:
I just realised that this section is not true anymore as I throw an error if the plugin was not found.

tipsy commented on src/main/java/io/javalin/core/JavalinConfig.java:
```diff
@@ -164,6 +202,24 @@ public void server(Supplier<Server> server) {
+            .filter(plugin -> plugin instanceof PluginLifecycleInit)
+            .forEach(plugin -> {
+                ((PluginLifecycleInit) plugin).init(app);
+                if (!registeredHandler.isEmpty() || !registeredWsHandler.isEmpty()) {
```
`applyUserConfig` happens once inside `create(...)`, won't these lists always be empty? Or is this for if the plugin itself is adding routes?

TobiasWalle replied:
> Or is this for if the plugin itself is adding routes?

Exactly! I want to detect this to guarantee that the `init` method is called before any handlers are registered.

TobiasWalle replied:
There is a test for this if you want an example.

tipsy replied:
Got it, looks good then!

tipsy commented:
I think this looks really good! I'll let @dherges have a look before I merge it.

tipsy commented on src/main/java/io/javalin/core/JavalinConfig.java:
```diff
@@ -164,6 +195,24 @@ public void server(Supplier<Server> server) {
 
     public static void applyUserConfig(Javalin app, JavalinConfig config, Consumer<JavalinConfig> userConfig) {
```
Side note: Maybe this method should be moved to Kotlin.

dherges commented:
looks good to me!
